### PR TITLE
Updated AsyncIndexLookup to wait for timed out threads to finish befo…

### PR DESCRIPTION
**Pull request by jwomeara**
…re accessing shared properties.

### CODE REVIEW

**Comment by billoley** [comment]
Changes look fine.  A possible enhancement since you're adding another parameter (CoundDownLatch) that gets passed from three different places would be to encapsulate the common parameters in a class that extends Future and contains the CountDownLatches and timeouts.

timedScanFuture aka Future<Boolean> that is passed along with two CountdDownLatches, startTimeMillis, and timeout could be enclosed in a class TimedScanFuture that would get created in BoundedRangeIndexLookup.createTimedCallable() and FieldNameIndexLookup.createTimedCallable() by submitting to the executor and returning the future there and in RegexIndexLookup.createTimedCallable() by doing the same thing and then storing the TimedScanFuture in RegexLookupData.

**Review by billoley — DISMISSED** [review]

**Comment by jwomeara** [comment]
> Changes look fine. A possible enhancement since you're adding another parameter (CoundDownLatch) that gets passed from three different places would be to encapsulate the common parameters in a class that extends Future and contains the CountDownLatches and timeouts.
> 
> timedScanFuture aka Future that is passed along with two CountdDownLatches, startTimeMillis, and timeout could be enclosed in a class TimedScanFuture that would get created in BoundedRangeIndexLookup.createTimedCallable() and FieldNameIndexLookup.createTimedCallable() by submitting to the executor and returning the future there and in RegexIndexLookup.createTimedCallable() by doing the same thing and then storing the TimedScanFuture in RegexLookupData.

So you're suggesting I create a wrapper around future then?  Not necessarily an extension of future, right?  Because the future is created by the executor on submit, so I can't control what type of future is created - i can only wrap what I receive.  

The timedWaitScan call doesn't have that many arguments IMO, and to me, passing the latches as arguments in the method in 3 places is no different than passing them to the constructor of whatever wraps the future in 3 places.  I would prefer to leave this as-is.

**Comment by billoley** [comment]
> So you're suggesting I create a wrapper around future then? Not necessarily an extension of future, right? Because the future is created by the executor on submit, so I can't control what type of future is created - i can only wrap what I receive.
> 
> The timedWaitScan call doesn't have that many arguments IMO, and to me, passing the latches as arguments in the method in 3 places is no different than passing them to the constructor of whatever wraps the future in 3 places. I would prefer to leave this as-is.

Yes, a wrapper (can still implement Future interface and encapsulate the returned Future) but I'm OK with leaving them as parameters.

File: warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java [hunk 1]
```diff
@@ -206,111 +207,116 @@ public synchronized IndexLookupMap lookup() {
...
+                            throw new Exception("Timeout exceeded for bounded range lookup");
+                        }
 
-                    k.getRow(holder);
-                    String uniqueTerm = holder.toString();
+                        Key k = entry.getKey();
 
-                    SortedMap<Key,Value> keymap = WholeRowIterator.decodeRow(entry.getKey(), entry.getValue());
+                        if (log.isTraceEnabled()) {
+                            log.trace("Foward Index entry: " + entry.getKey());
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java]: small typo here: Forward Index entry:

File: warehouse/query-core/src/main/java/datawave/query/jexl/lookups/RegexIndexLookup.java [hunk 2]
```diff
@@ -251,107 +251,113 @@ public synchronized IndexLookupMap lookup() {
...
-                        log.trace("Foward Index entry: " + entry.getKey());
-                    }
+                        Key topKey = entry.getKey();
 
-                    // Get the column qualifier from the key. It contains the datatype and normalizer class
-                    if (null != topKey.getColumnQualifier()) {
-                        String colq = topKey.getColumnQualifier().toString();
-                        int idx = colq.indexOf(Constants.NULL);
+                        if (log.isTraceEnabled()) {
+                            log.trace("Foward Index entry: " + entry.getKey());
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/jexl/lookups/RegexIndexLookup.java]: Similar typo here.

**Review by keith-ratcliffe — APPROVED** [review]
The changes here look good to me

**Review by hlgp — APPROVED** [review]

**Pull request merged**